docs/data/data-grid/server-side-data/lazy-loading.md

```diff
@@ -2,14 +2,107 @@
 title: React Server-side lazy loading
 ---
 
-# Data Grid - Server-side lazy loading [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')🚧
+# Data Grid - Server-side lazy loading [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')
 
 <p class="description">Row lazy-loading with server-side data source.</p>
 
+Lazy Loading changes the way pagination works by removing page controls and loading data dynamically (in a single list) as the user scrolls through the grid.
+
+It is enabled by adding `lazyLoading` prop in combination with `unstable_dataSource` prop.
+
+Initially, the first page data is fetched and displayed in the grid. What triggers the loading of next page data depends on the value of the total row count.
+
+If the total row count is known, the grid gets filled with skeleton rows and fetches more data if one of the skeleton rows falls into the rendering context.
+
+If the total row count is unknown, the grid fetches more data when the user scrolls to the bottom of the grid. This loading strategy is often referred to as **infinite loading**.
+
+:::info
+Row count can be provided either by returning the `rowCount` in the response of the `getRows` method in `unstable_dataSource`, via the `rowCount` prop or by calling [`setRowCount`](/x/api/data-grid/grid-api/#grid-api-prop-setRowCount) API.
+:::
+
+:::warning
+Order of precedence for the row count:
+
+- `rowCount` prop
+- `rowCount` returned by the `getRows` method
+- row count set using the `setRowCount` API
+
+This means that, if the row count is set using the API, that value gets overridden once a new value is returned by the `getRows` method, even if it is `undefined`.
```

Comment:
> even if it is `undefined`

Does that mean that even if there's no `rowCount` returned from the `getRows` call (equivalent to being `undefined`), the previous value set using the API method will be replaced?

Could that be against some users' expectations? 🤔

Reply:
The problem is, we use `undefined` to say that you don't know the row count. That switches the grid into infinite loading mode.
If you set the value trough API and then make another call and give back undefined as `rowCount`, does it mean
1. Row count became unknown
2. Do not update row count

RIght now, it works as point 1.
If you know row count, but it is not returned by the response, you can always do this
```
getRows: async (params: GridGetRowsParams) => {
  const urlParams = new URLSearchParams({
    ...
  });
  const getRowsResponse = await fetchRows(
    ...
  );

  return {
    rows: getRowsResponse.rows,
    rowCount: myCountVar
  };
},
```

I see setting the page count through API more as way to postpone count and update it after the data is fetched. Ideally, your server should return it with the data

Reply:
There could be cases when the `rowCount` is not accompanied by the regular server call to fetch a page, e.g. when it's expensive on the server to compute row count (Imagine data stored as a linked list, it's needed to be completely traversed to know actual `rowCount`)
In those cases, it should be possible to skip `rowCount` update on each `getRows` call, rather the user could separately fetch it and set using the imperative API.

What do you think about only treating `-1` as an unknown row count?

It will also be consistent with the [client side pagination](https://mui.com/x/react-data-grid/pagination/#index-based-pagination) which explicitly requires `rowCount=-1` to consider the grid in unknown or estimated row count mode.

Reply:
Updated.
New condition for the row update
`response.rowCount !== undefined || pageRowCount === undefined`
not passing rowCount while it is not even set, results in setting it to `-1`